src/test/mod.rs

```diff
@@ -2,13 +2,55 @@ mod fakes;
 
 use crate::web::Server;
 use failure::Error;
+use once_cell::unsync::OnceCell;
 use postgres::Connection;
-use std::sync::{Arc, Mutex, MutexGuard};
 use reqwest::{Client, Method, RequestBuilder};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+pub(crate) fn wrapper(f: impl FnOnce(&TestEnvironment) -> Result<(), Error>) {
+    let env = TestEnvironment::new();
+    let result = f(&env);
+    env.cleanup();
+
+    if let Err(err) = result {
+        eprintln!("the test failed: {}", err);
+        for cause in err.iter_causes() {
+            eprintln!("  caused by: {}", cause);
+        }
+
+        eprintln!("{}", err.backtrace());
 
-pub(crate) fn with_database(f: impl FnOnce(&TestDatabase) -> Result<(), Error>) {
-    let env = TestDatabase::new().expect("failed to initialize the environment");
-    report_error(|| f(&env));
+        panic!("the test failed");
```

Comment:
Why not `panic!("the test failed: {}", err)`?

Reply:
Because the error message is already printed above, along with the causes and the backtrace when available: the panic is only needed to mark the test as failed.

Reply:
Right, my question is why not replace the 5 or so lines of boilerplate with `panic!` ?

Reply:
Because panic doesn't show the causes of the error, nor the error backtrace (it shows instead the `panic!()` backtrace).

Reply:
Ah right, I forgot `Error` has its own backtrace.